src/python_testing/TC_TLSCERT.py

```diff
@@ -275,295 +290,446 @@ async def read_tls_cert_attribute(self, attribute):
-            asserts.assert_equal(found_certs[response2.caid], second_cert, "Expected matching certificate detail")
+        await cr2_cmd.send_provision_root_command(certificate=my_root_cert[2], caid=my_caid[0], expected_status=Status.NotFound)
 
         self.step(15)
-        await cr1_cmd.send_remove_root_command(caid=response.caid)
-        if not self.skip_cr2:
-            await cr2_cmd.send_remove_root_command(caid=response2.caid)
-        resp = await self.send_single_cmd(cmd=Clusters.OperationalCredentials.Commands.RemoveFabric(fabric_index_cr2), endpoint=0)
-        asserts.assert_equal(
-            resp.statusCode, Clusters.OperationalCredentials.Enums.NodeOperationalCertStatusEnum.kOk)
-
-    def desc_TC_TLSCERT_3_1_8(self) -> str:
+        response = await cr2_cmd.send_provision_root_command(certificate=my_root_cert[2])
+        self.assert_valid_caid(response.caid)
+        my_caid[1] = response.caid
+
+        self.step(16)
+        await cr2_cmd.send_provision_root_command(certificate=my_root_cert[2], caid=my_caid[1], expected_status=Status.AlreadyExists)
+
+        self.step(17)
+        found_certs = await cr1_cmd.read_root_certs_attribute_as_map()
+        asserts.assert_in(my_caid[0], found_certs, "ProvisionedRootCertificates should contain provisioned root cert")
+        asserts.assert_equal(len(found_certs), 1, "Expected 1 certificate")
+        asserts.assert_equal(found_certs[my_caid[0]].certificate, my_root_cert[1], "Expected matching certificate detail")
+
+        self.step(18)
+        found_certs = await cr2_cmd.read_root_certs_attribute_as_map()
+        asserts.assert_in(my_caid[1], found_certs, "ProvisionedRootCertificates should contain provisioned root cert")
+        asserts.assert_equal(len(found_certs), 1, "Expected 1 certificate")
+        asserts.assert_equal(found_certs[my_caid[1]].certificate, my_root_cert[2], "Expected matching certificate detail")
+
+        self.step(19)
+        find_response = await cr1_cmd.send_find_root_command()
+        asserts.assert_equal(len(find_response.certificateDetails), 1, "Expected single certificate")
+        asserts.assert_equal(find_response.certificateDetails[0].certificate, my_root_cert[1])
+        asserts.assert_equal(find_response.certificateDetails[0].caid, my_caid[0])
+
+        self.step(20)
+        find_response = await cr2_cmd.send_find_root_command()
+        asserts.assert_equal(len(find_response.certificateDetails), 1, "Expected single certificate")
+        asserts.assert_equal(find_response.certificateDetails[0].certificate, my_root_cert[2])
+        asserts.assert_equal(find_response.certificateDetails[0].caid, my_caid[1])
+
+        self.step(21)
+        await cr1_cmd.send_remove_root_command(caid=my_caid[0])
+
+        self.step(22)
+        await cr1_cmd.send_remove_fabric_command(cr2_cmd.fabric_index)
+
+    def desc_TC_TLSCERT_2_9(self) -> str:
```

Comment:
suggest adding pics_ functions for each of the tests too. I know they're marked as only applicable for devices with that cluster on that endpoint, which is actually really helpful for when we start batch-running tests, but the TH still goes by PICS for test selection, so it'll select these tests for every device, then force them to run the test and see that it gets skipped. I know...not ideal. Getting rid of PICS has been a slog and a rant for another day. Anyway, doesn't need to be done in the PR or before the test event, but should add these to give some guidance to the TH.

Reply:
Added, hopefully got the right ones (thanks!), guessed at the `.S` prefix as it seemed to be what everything did